First draft of multi-tenancy proposal (#1149)
kubernetes-sigs/cluster-api-provider-vsphere: docs/proposal/20210311-single-controller-multitenancy.md

        - [Current State](#current-state)
        - [Proposed Changes](#proposed-changes)
        - [Controller Changes](#controller-changes)
        - [Clusterctl Changes](#clusterctl-changes)
    - [Security Model](#security-model)
        - [Roles](#roles)
        - [RBAC](#rbac)
        - [Namespace Restrictions](#namespace-restrictions)
        - [CAPV Controller Requirements](#capv-controller-requirements)
    - [Risks and Mitigations](#risks-and-mitigations)
        - [Caching](#caching)
  - [Alternatives](#alternatives)
    - [Using Secrets](#using-only-secrets-to-specify-vsphere-accounts)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
  - [Implementation History](#implementation-history)

## Glossary
- CAPV - An abbreviation of Cluster API Provider vSphere

## Summary
The CAPV controller is capable of managing infrastructure resources on a vCenter using the credentials it was provided during initialization. The credentials are provided via environment variables that get saved onto a secret that's used by the CAPV deployment. 

> Credentials provided are used for the entire lifetime of the CAPV deployment which means a CAPV cluster can become broken if the provisioning CAPV deployment were to be reconfigred for another set of credentials.

Reviewer: How multitenancy can fix CAPV deployment credential issue? 
CAPA will start with credentials retrieved locally, then other roles could be used for creating workload clusters. 
When the local credentials (that CAPV uses) change, is the plan to update the secret?

Author: I don't think MT can fix any existing clusters from breaking when the deployment credentials are changed. This was more highlighting a cluster's dependance on the deployment credentials and if MT is fully utilized, it can move that dependency to a `VSphereAccount`

<strong><em>Changes to VSphereCluster</strong></em>
A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.

```go
// VSphereAccount is the account to be used for vcenter actions
type VSphereAccount struct {
    metav1.TypeMeta `json:`",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec VSphereAccountSpec `json:"spec,omitempty"`
    Status VSphereAccountStatus `json:"status, omitempty"`
}

type VSphereAccountSpec struct {
    // Server is the address of the vSphere endpoint.
    Server string `json:"server,omitempty"`

    Username string `json:"username,omitempty"`

    Password string `json:"password,omitempty"`

    // AllowedNamespaces is an array of namespaces that VSphereClusters can
    // use this account from.
    //
    // An empty list (default) indicates that VSphereClusters can use this
    // account from any namespace. This field is intentionally not a
    // pointer because the nil behavior (no namespaces) is undesirable here.
    // +optional

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer 2: @sedefsavas we had some issues with implementing the label selector and that's why we only did the []string, but I do agree its good a consistent behaviour for the default across providers

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Author: I like both the differences with CAPA. Was originally thinking label selector was overkill but I can see how that's useful with targeting multiple namespaces without having to list each of them out. Will update, thanks Sedef

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer 3: +1 to consistent behavior across providers.

What is the behavior when `NamespaceList` is nil and a selector is set? Is the result always the union of `NamespaceList` + `Selector`?

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Author: yeah in CAPA's implementation, it is a union of the two. That'll offer the most flexibility

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer: empty `allowedNamespaces` --> allows all namespaces 
nil `allowedNamespaces` --> blocks all namespaces 

nil `NamespaceList` --> blocks all
nil `Selector` --> blocks all

> What is the behavior when NamespaceList is nil and a selector is set? Is the result always the union of NamespaceList + Selector?
Yes, if at least one of them matches a namespace, it is allowed.

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer: I drafted some examples for how we use allowedNamespaces in CAPA:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2319/files

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer 4: I'm tempted to say we pull out the list option for v1alpha4. I will take the conversation to CAPI, but https://github.com/kubernetes/kubernetes/pull/101342 is going to move NamespaceDefaultLabelName to GA and has been defaulted to on from 1.21 .

>     AllowedNamespaces []string `json:"allowedNamespaces"`

Reviewer 4: PTAL https://github.com/kubernetes-sigs/cluster-api/pull/4514

# Single Controller Multitenancy

```text
---
title: Single Controller Multitenancy
authors:
  - "@gab-satchi"

> reviewers:

Author: ACK

##### FR2. CAPV MUST support static credentials

##### FR3. CAPV MUST support clusterctl move scenarios

##### FR4. CAPV MUST prevent privilege escalation allowing users to create clusters in accounts they should not be able to

#### Non-Functional Requirements

##### NFR1. Unit tests MUST exist for cluster and machine controllers that utilize the credentials

##### NFR2. e2e tests MUST exist for multi account scenarios

### Implementation Details/Notes/Constraints

#### Current State

CAPV currently uses a session manager to create and cache sessions. Sessions are created (or retrieved from the cache) during cluster and vsphere VM reconcile loops.
vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.

```go
type VMContext struct {
    *ControllerContext
    VSphereVM   *v1alpha3.VSphereVM
    PatchHelper *patch.Helper
    Logger      logr.Logger
    Session     *session.Session
}
```

#### Proposed Changes

The proposed changes below allow users to specify the account to use for clusters during runtime.

Changed Resources

- `VSphereCluster`

New Resources

> - cluster-scoped `VSphereAccount` represents the vCenter account to use for reconciliation. This type should also contain a string list of namespaces in which it is allowed to be used.

Author: It's called out a bit in `Clusterctl Changes`. I don't think there's anything concrete in clusterctl regarding that yet. Some discussion captured here: https://github.com/kubernetes-sigs/cluster-api/issues/3354

We'd likely do something where we copy accounts (while avoiding duplicates) and almost never delete accounts as it can be used by other clusters or intended to be used for future clusters.

Changes to VSphereCluster
A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.

```go
// VSphereAccount is the account to be used for vcenter actions
type VSphereAccount struct {
    metav1.TypeMeta `json:`",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec VSphereAccountSpec `json:"spec,omitempty"`
    Status VSphereAccountStatus `json:"status, omitempty"`
}

type VSphereAccountSpec struct {
    // Server is the address of the vSphere endpoint.
    Server string `json:"server,omitempty"`

>     Username string `json:"username,omitempty"`

Reviewer 5: remove `omitempty` ?

>     Password string `json:"password,omitempty"`

Reviewer 5: same

    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
    // Namespaces can be selected either by using an array of namespaces or with a label selector.

>     // If this object is nil, no namespaces will be allowed

Reviewer 5: are other providers failing closed too ?

Author: AWS is but Azure uses a string slice and if it's empty, will default to all namespaces are allowed.

Reviewer 2: I'm keep track of this to discuss with capz community

Reviewer 5: @randomvariable @sedefsavas -  IIRC CAPA also has namespaced resources for account management right ? so that might be a stop-gap until clusterctl supports the cluster-scoped ones

Reviewer 4: Yes, the idea that you can enable both the cloud admin persona to set up cluster scoped resources for use by different teams, and then the devops and developer personas can also plug in their own dev environments using the namespace scoped resources.

Reviewer 5: @gab-satchi - this should be a stop-gap until we know how to move cluster-scoped resources

Changes to VSphereCluster
A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.

```go
// VSphereAccount is the account to be used for vcenter actions
type VSphereAccount struct {
    metav1.TypeMeta `json:`",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec VSphereAccountSpec `json:"spec,omitempty"`
    Status VSphereAccountStatus `json:"status, omitempty"`
}

type VSphereAccountSpec struct {
    Username string `json:"username"`

    Password string `json:"password"`

    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
    // Namespaces can be selected either by using an array of namespaces or with a label selector.
    // If this object is nil, no namespaces will be allowed
    //
    // +optional
    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
}

type AllowedNamespaces struct {

>     // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)

Reviewer 6: What if both are specified? How do they combine (or or AND)?

Author: It will be the union so OR. Will update to make that clearer

    //
    // Allowed namespaces listed out by name
    // +optional
    // +nullable
    NamespaceList []string `json:"list"`

    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
    // +optional
    Selector metav1.LabelSelector `json:"selector"`
}

type VSphereClusterSpec struct {
    ...
    // +optional
    AccountRef *corev1.LocalObjectReference `json:"accountRef,omitempty"`
}
```

> #### Controller Changes

Author: updated Controller Changes to include suggestions.

> - 1If AccountRef is specified, the CRD is fetched and used to create a session

Reviewer 6: ```suggestion
- If AccountRef is specified, the CRD is fetched and used to create a session
```

- The session will be cached using the existing logic where the key is server + username + datacenter
- The controller should reconcile the VSphereAccount resource and ensure connectivity to the account

> - The controller will set the VsphereAccount as one of the OwnerReferences of the VsphereCluster if the VsphereAccount is in the same namespace as the VsphereCluster

Reviewer 6: I'm a little bit confused by this, my understanding till now was that VsphereAccount is a cluster scoped resource...

Author: It should be cluster scoped. CAPA called out cluster scoped vs. namespaced in the proposal but the implementation (so far) uses cluster scoped identities and doesn't set the owner reference. CAPZ seems to use namespaced crds for the AzureIdentity but also copies the identity per cluster.

#### Clusterctl Changes

> Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track ownerReferences pointing at cluster-scoped resources, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces.

Reviewer 6: > clusterctl's object graph to track ownerReferences pointing at cluster-scoped resources

As far as I understood, the only link to VsphereAccount is AccountRef, and this makes things a little bit more complicated because the object tree is defined outside the ownerRefereces. 
We can define a solution to address this, but I really would like to come out with a design that works for all the providers, not only for CAPV
 
Also, what is the expected behavior if the cluster scope resource already exists in the target cluster? (override or ignore)

Reviewer 4: >  the more it makes sense to have a 1:1 relationship with a cluster and account
I think this will defeat the purpose somewhat, as it limits the personas who can create clusters.

#### Non-Functional Requirements

##### NFR1. Unit tests MUST exist for cluster and machine controllers that utilize the credentials

##### NFR2. e2e tests MUST exist for multi account scenarios

### Implementation Details/Notes/Constraints

#### Current State

CAPV currently uses a session manager to create and cache sessions. Sessions are created (or retrieved from the cache) during cluster and vsphere VM reconcile loops.
vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.

```go
type VMContext struct {
    *ControllerContext
    VSphereVM   *v1alpha3.VSphereVM
    PatchHelper *patch.Helper
    Logger      logr.Logger
    Session     *session.Session
}
```

#### Proposed Changes

The proposed changes below allow users to specify the account to use for clusters during runtime.

Changed Resources

- `VSphereCluster`

New Resources

- cluster-scoped `VSphereAccount` represents the vCenter account to use for reconciliation. This type should also contain a string list of namespaces in which it is allowed to be used.

Changes to VSphereCluster

> A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.

Reviewer 4: We've created our own PrincipalReference type in CAPA rather than reuse TypedLocalObjectReference.

```go
// VSphereAccount is the account to be used for vcenter actions

> type VSphereAccount struct {

Reviewer 4: ```suggestion
type VSphereClusterAccount struct {
```

Cluster scoped resources should have "Cluster" in the name.

    metav1.TypeMeta `json:`",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec VSphereAccountSpec `json:"spec,omitempty"`
    Status VSphereAccountStatus `json:"status, omitempty"`
}

type VSphereAccountSpec struct {
    Username string `json:"username"`

>     Password string `json:"password"`

Reviewer 7: shall we put this username/password inside a secret?

Reviewer 4: Yes, the actual credential should still be a SecretRef

```go
// VSphereClusterIdentity is the account to be used for vcenter actions
type VSphereClusterIdentity struct {
    metav1.TypeMeta `json:`",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec VSphereClusterIdentitySpec `json:"spec,omitempty"`
    Status VSphereClusterIdentityStatus `json:"status, omitempty"`
}

type VSphereClusterIdentitySpec struct {
    // secretsRef references a secret in the CAPV controller namespace with the credentials to use
    secretRef `json:"secretRef,omitempty"`

    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
    // Namespaces can be selected either by using a label selector.
    // If this object is nil, no namespaces will be allowed.
    // If this object is empty {}, clusters can use this identity from any namespace.
    //
    // +optional
    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
}

> type AllowedNamespaces struct {

Reviewer 6: according the general guidelines being proposed in CAPI, NamespaceList is missing

Reviewer 5: +1

Author: The PR calls out to deprecate `NamespaceList` for v1alpha4

`Providers SHOULD mark NamespaceList as deprecated from v1alpha4`

    // Selector is a standard Kubernetes LabelSelector.
    // This is a standard Kubernetes LabelSelector, a label query over a set of resources.
    // The result of matchLabels and matchExpressions are ANDed.
    // +optional
    Selector metav1.LabelSelector `json:"selector"`
}

type VSphereIdentityReference struct {
    // Kind of the identity.
    // +kubebuilder:validation:Enum=VSphereClusterIdentity,Secret
    Kind string

    // Name of the identity
    // +kubebuilder:validation:MinLength=1
    Name string
}

type VSphereClusterSpec struct {
    ...
    // +optional
    IdentityRef *VSphereIdentityReference `json:"IdentityRef,omitempty"`
}
```

#### Controller Changes

* If IdentityRef is specified, and it references a `VSphereClusterIdentity`, the CRD is fetched and used to create a session
* The secret referenced within `VSphereClusterIdentity` must be in the controller namespace

> * If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session. The secret must be in the same namespace as the `VSphereCluster`

Reviewer 6: > The secret must be in the same namespace as the `VSphereCluster`

Am I wrong or the secrets for Global resources should be in controller namespace?

Reviewer 5: this is for the local credentials (i.e. BYO creds within the namespace where you provision your cluster). But I agree for the global ones

Author: yeah when the secret is addressed through the global `VSphereClusterIdentity` then it's expected to be in the controller namespace. CAPV is also implementing a direct reference to a secret from a `VSphereCluster` and for that we expect it to be in the same namespace as the cluster.

* The controller will add the VSphereCluster to OwnerReferences on a Secret that contains the account credentials
* If IdentityRef is not specified, the controller will fallback to using credentials provided in the controller deployment
* The session will be cached using the existing logic where the key is server + username + datacenter
* The `IdentityRef` on a `VSphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
* The controller will fail to delete a `VSphereClusterIdentity` with an error if the account is used by `VSphereClusters`

#### Clusterctl Changes

> Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VSphereClusterIdentity` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.

Reviewer 6: > Any `VSphereClusterIdentity` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.

Is this still a requirement? AFAIK there is no need for manually moving objects (except you explicitly wont to do so)

Reviewer 5: agreed

Author: thanks will update.

```go
type VMContext struct {
    *ControllerContext
    VSphereVM   *v1alpha3.VSphereVM
    PatchHelper *patch.Helper
    Logger      logr.Logger
    Session     *session.Session
}
```

#### Proposed Changes

The proposed changes below allow users to specify the account to use for clusters during runtime. The credentials can be provided by two approaches:

- Referencing a cluster scoped `VsphereClusterAccount` to be used for credentials
- Referencing a `Secret` that contains the credentials

Changed Resources

- `VsphereCluster`

New Resources

- cluster-scoped `VsphereClusterAccount` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.

Changes to VsphereCluster
A new field is added to the VsphereClusterSpec to reference the `VsphereClusterAccount`. We intend to use `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.

```go
// VsphereClusterAccount is the account to be used for vcenter actions

> type VsphereClusterAccount struct {

Reviewer 6: I know that naming is being discussed with other providers, so feel to ignore…
Given that this type is referenced by `identityRef`, should this be `VsphereClusterIdentity` (or simply `VSphereIdentity`)?

    CredentialsRef *corev1.ObjectReference `json:"credentialsRef,omitempty"`

    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
    // Namespaces can be selected either by using an array of namespaces or with a label selector.
    // If this object is nil, no namespaces will be allowed
    //
    // +optional
    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
}

type AllowedNamespaces struct {
    // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)
    // If both NamespaceList and Selector are defined, the union of the two would be used to determine if a namespace is allowed to use the account
    //
    // Allowed namespaces listed out by name
    // +optional
    // +nullable
    NamespaceList []string `json:"list"`

    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
    // +optional
    Selector metav1.LabelSelector `json:"selector"`
}

type VsphereClusterSpec struct {
    ...
    // +optional
    IdentityRef *corev1.TypedLocalObjectReference `json:"IdentityRef,omitempty"`
}
```

#### Controller Changes

> - If IdentityRef is specified, and it references a `VsphereClusterAccount`, the CRD is fetched and used to create a session

Reviewer 5: @randomvariable @fabriziopandini  -  it might worthwhile to inject this reference if nothing is set **and** there are `VsphereClusterAccount` that are allowed for that namespace. This avoids having to communicate the kind and Name to folks that want to provision clusters and that don't care much about creds ?

Reviewer 6: I find the idea interesting, because it makes it easier to decouple roles and responsibilities
What there are more than one Account eligible for a cluster?

Reviewer 4: We have a mechanism in CAPA for that called AWSClusterControllerIdentity that respects the existing behaviour as well as EC2 instance profiles for cross-account role assumption which is a key requirement for AWS. See https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/proposal/20200506-single-controller-multitenancy.md#upgrade-strategy

Azure doesn't have cross-account role assumption so this wasn't an important use case for them, so was not implemented there.

> - If IdentityRef is specified, and it references a `VsphereClusterAccount`, the CRD is fetched and used to create a session

Reviewer 4: ```suggestion
- If IdentityRef is specified, and it references a `VSphereClusterAccount`, the CRD is fetched and used to create a session
```

- If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session

> - The controller will add the VsphereCluster to OwnerReferences on a Secret that contains the account credentials

Reviewer 4: ```suggestion
- The controller will add the VSphereCluster to OwnerReferences on a Secret that contains the account credentials
```

- If IdentityRef is not specified, the controller will fallback to using credentials provided in the controller deployment
- The session will be cached using the existing logic where the key is server + username + datacenter

> - The `IdentityRef` on a `VsphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials

Reviewer 4: ```suggestion
- The `IdentityRef` on a `VSphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
```

Reviewer 4: Consider https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/proposal/20200506-single-controller-multitenancy.md#upgrade-strategy for fallback behaviour.

> - The controller will fail to delete a `VsphereClusterAccount` with an error if the account is used by `VsphereClusters`

Reviewer 4: ```suggestion
- The controller will fail to delete a `VSphereClusterAccount` with an error if the account is used by `VSphereClusters`
```

> #### Clusterctl Changes

Reviewer 6: Might be I'm wrong, but I see a potential problem for move in the fact that `VsphereClusterAccount` refers to a secret, and this secret can be in any namespace

- Say we are moving clusters in namespace A
- With the suggested changes, move will include also `VsphereClusterAccount` objects
- Say that a VsphereClusterAccount store its credential in the namespace B 

As clusterctl is designed today (https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#move), secret in namespace B won't be moved thus leading to an inconsistent `VsphereClusterAccount` in the target cluster...

Reviewer 4: Cluster scoped resources should not support arbitrary namespaces for secrets IMO, but only the controller's own namespace, that is what I've put in https://github.com/kubernetes-sigs/cluster-api/pull/4514/files

> Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VsphereClusterAccount` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.

Reviewer 6: > clusterctl's object graph to track cluster-scoped resources that are used by the source cluster,

This is kind of complex, given that the only relation between cluster and `VsphereClusterAccount` is defined in the identityRef field, and thus unknown to the clusterctl move discovery algorithm (that should be generic/provider agnostic).

There are two possible workarounds: 
- Move all the `VsphereClusterAccount` (by adding the force move label on the VsphereClusterAccount  CRD)
- Move only specific `VsphereClusterAccount` (by adding the force move label only on those objects)

> Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VsphereClusterAccount` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.

Reviewer 4: ```suggestion
Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VSphereClusterAccount` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.
```

  application permissions.

- **Workload Cluster Operator**: The workload cluster operator (dev) is responsible for
  management of the cluster relevant to their particular applications .

There are two primary components to the Service APIs security model: RBAC and namespace restrictions.

#### RBAC

RBAC (role-based access control) is the standard used for Kubernetes
authorization. This allows users to configure who can perform actions on
resources in specific scopes. RBAC can be used to enable each of the roles
defined above. In most cases, it will be desirable to have all resources be
readable by most roles, so instead we'll focus on write access for this model.

##### Write Permissions

|                              | VsphereClusterAccount | Secret | Cluster |
| ---------------------------- | --------------------- | ------ | ------- |
| Infrastructure Provider      | Yes                   | Yes    | Yes     |
| Management Cluster Operators | Yes                   | Yes    | Yes     |
| Workload Cluster Operator    | No                    | Yes    | Yes     |

#### Namespace Restrictions

> - To prevent workload cluster operators from using cluster-scoped `VsphereClusterAccounts` that they should not be using, the `allowedNamespaces` will be used to dictate which namespaces are allowed to use the `VsphereClusterAccount`.

Reviewer 4: ```suggestion
- To prevent workload cluster operators from using cluster-scoped `VSphereClusterAccounts` that they should not be using, the `allowedNamespaces` will be used to dictate which namespaces are allowed to use the `VSphereClusterAccount`.
```

An empty slice indicates that the `VsphereClusterAccount` can be used by any `VsphereCluster`

#### CAPV Controller Requirements

The CAPV controller will need to:

> - Populate conditions when cluster is misconfigured. `VsphereClusterAccount` is not found or isn't compatible with the cluster due to namespace restrictions.

Reviewer 4: ```suggestion
- Populate conditions when cluster is misconfigured. `VSphereClusterAccount` is not found or isn't compatible with the cluster due to namespace restrictions.
```

- Not implement an invalid configuration. If a cluster is attempting to use the `VsphereClusterAccount` from an invalid namespace, ignore it and indicate the issue through conditions.

> - Respond to changes in a `VsphereClusterAccount` spec.

Reviewer 4: ```suggestion
- Respond to changes in a `VSphereClusterAccount` spec.
```

### Risks and Mitigations

#### Caching

With multi-tenant support, a single CAPV instance may reconcile multiple clusters across many accounts.
There's existing caching in the session manager that should be sufficient to handle the extra sessions that will get created.

## Alternatives

### Using only secrets to specify vSphere accounts

The VsphereCluster will reference the Secret via an ObjectReference

#### Benefits

- Re-using secrets ensures encryption by default and provides a clear UX signal to end users that the data is meant to be secure
- Keeps clusterctl move straightforward with the 1:1 cluster -> credential relationship

#### Mitigations for current proposal

> - There are usecases where users would like to reuse the same account for multiple namespaces. See [Story 2](#story-2---deploying-multiple-clusters-from-a-single-account)

Reviewer 4: ```suggestion
- There are use cases where users would like to reuse the same account for multiple namespaces. See [Story 2](#story-2---deploying-multiple-clusters-from-a-single-account)
```
